sdk/servicebus/service-bus/src/util/utils.ts

```diff
@@ -485,3 +487,96 @@ export type EntityStatus =
   | "Renaming"
   | "Restoring"
   | "Unknown";
+
+/**
+ * An executor for a Promise (or function that returns a Promise) that obeys both a timeout and an
+ * optional AbortSignal.
+ * @param timeoutMs - The number of milliseconds to allow before throwing an AMQP error with condition ServiceUnavailableError
+ * @param timeoutMessage - The message to place in the .description field for the thrown exception for Timeout.
+ * @param abortSignal - The abortSignal associated with containing operation.
+ * @param abortErrorMsg - The abort error message associated with containing operation.
+ * @param value - The value to be resolved with after a timeout of t milliseconds.
+ * @returns {Promise<T>} - Resolved promise
+ *
+ * @internal
+ * @ignore
+ */
+export function waitForTimeoutOrAbortOrResolve<T>(args: {
+  actionFn: () => Promise<T>;
+  timeoutMs: number;
+  timeoutMessage: string;
+  abortSignal?: AbortSignalLike;
+  abortMessage?: string;
+}): Promise<T> {
+  if (args.abortSignal && args.abortSignal.aborted) {
+    throw new AbortError(args.abortMessage);
+  }
+
+  let timer: any | undefined = undefined;
+  let clearAbortSignal: (() => void) | undefined = undefined;
+
+  const clearAbortSignalAndTimer = (): void => {
+    clearTimeout(timer);
+
+    if (clearAbortSignal) {
+      clearAbortSignal();
+    }
+  };
+
+  const abortOrTimeoutPromise = new Promise((resolve, reject) => {
+    clearAbortSignal = checkAndRegisterWithAbortSignal(reject, args.abortMessage, args.abortSignal);
+
+    // using a named function here so we can identify it in our unit tests
+    timer = setTimeout(function timeoutCallback() {
+      reject({
+        condition: ErrorNameConditionMapper.ServiceUnavailableError,
+        description: args.timeoutMessage
+      } as AmqpError);
+    }, args.timeoutMs);
+  });
+
+  const actionPromise = args.actionFn();
+  return Promise.race([abortOrTimeoutPromise, actionPromise])
+    .then(() => {
+      clearAbortSignalAndTimer();
+      return actionPromise;
+    })
+    .catch((err) => {
+      clearAbortSignalAndTimer();
+      throw err;
+    });
+}
+
+/**
+ * Checks the abort signal to see if it's already aborted (and if so, throws an AbortError).
+ *
+ * If it is not signalled, adds an event listener that will call your abortFn.
+ *
+ * @returns A function that removes any of our attached event listeners on the abort signal or an empty function if
+ * the abortSignal was not defined.
+ *
+ * @internal
+ * @ignore
+ */
+export function checkAndRegisterWithAbortSignal(
+  onAbortFn: (abortError: AbortError) => void,
+  abortMessage?: string,
```

Comment:
(https://github.com/Azure/azure-sdk-for-js/pull/8994#discussion_r430891849)

Ramya:
> In which case would one call checkAndRegisterWithAbortSignal but not pass in the error message to use?

Richard:
> If you just want an AbortError and don't care about the message. It's optional in the AbortError as well

Ramya:
> It being optional in AbortError might have been due to us modeling it after Error which is a class we can instantiate without error message. But, from the Service Bus client perspective, I am not aware of a scenario where we would not want an accompanying error message. If so, then can we make it mandatory?

Reply:
Makes sense. I'll plumb that through.